Add support for using client count, and fixed input delay, for all inputs.

joeldenning opened the pull request:
The goal of the `use_client_count` option is to support only using the inputs sent by the client on the frame sent by the client, so that network latency doesn't change which frame an input occurs on.

I believe that this is the only change needed to ensure a consistent input delay (with the exception of lag) during netplay. From reading the code, I think that the `buffer_size` won't change which inputs are assigned to which frames, due to my changes in the checkifExists function. However, I'm not sure about that and gladly welcome your feedback.

joeldenning commented on socketServer.cpp:
```diff
@@ -101,7 +101,8 @@ void SocketServer::processBinaryMessage(QByteArray message)
                     writeLog(QString("creating ") + (lle ? "LLE" : "HLE") + " room", json.value("room_name").toString(), json.value("game_name").toString(), port);
 
-                    ServerThread *serverThread = new ServerThread(port, this);
+                    bool useClientCount = json.contains("use_client_count") ? json.value("use_client_count").toBool() : false;
```
I have a corresponding change to the mupen64plus-gui that sets this variable based on an option in the Create Room form.

joeldenning commented on serverThread.h:
```diff
@@ -7,7 +7,7 @@ class ServerThread : public QThread
     void run() Q_DECL_OVERRIDE;
 public:
-    ServerThread(int _port, QObject *parent = 0);
+    ServerThread(int _port, QObject *parent = 0, bool useClientCount = false);
```
The `useClientCount` option is meant to be backwards compatible, so it defaults to false.

loganmc10 commented:
This is pretty close, but I can see one issue:
```
    checkIfExists(playerNum, count);
    qToBigEndian(inputs[playerNum].value(count).first, &buffer[curr]);
    curr += 4;
    buffer[curr] = inputs[playerNum].value(count).second;
```
The ```checkIfExists``` function is probably poorly named, it's more like "makeItExist". With the changes you made, you need to account for the situation where somebody requests input, but the server doesn't have it. In this case, ```inputs[playerNum].value(count)``` is going to be empty.

You probably don't run into this situation when you play locally, but it'll happen over the Internet.

I believe the correct thing to do would be to just return early from ```sendInput()``` without sending anything to the client, the client will retry again later, that's how it's designed

loganmc10 commented:
When you bundle this with your input delay change, you'll probably run into the issue where the first few inputs are unknown (it will hang).

Like it needs inputs for count 0 and 1, but you only start sending inputs for count 2 and above, hopefully that makes sense? You'll need some way to fill in those first early frames with 0's. Basically if you set the input delay to 2, you just need to make the server fill in counts 0 and 1 with 0's for all players

joeldenning commented:
Thanks for the review - I'll push updates in the next few days.

joeldenning commented on serverThread.cpp:
```diff
@@ -2,11 +2,12 @@
 #include "tcpServer.h"
 
-ServerThread::ServerThread(int _port, QObject *parent)
+ServerThread::ServerThread(int _port, QObject *parent, int _p1InputDelay)
```
I'm passing in player 1's input delay in the constructor, but players 2-4 via the `setInputDelay` functions. The reason is that I found that there was a race condition with the server thread starting and the create_room code emitting the inputDelayChanged signal.

joeldenning commented on socketServer.cpp:
```diff
@@ -201,6 +204,10 @@ void SocketServer::processBinaryMessage(QByteArray message)
             clients[room_port].append(qMakePair(client, qMakePair(json.value("player_name").toString(), player_num)));
+
+            if (json.contains("input_delay")) {
+                emit inputDelayChanged(player_num - 1, json.value("input_delay").toInt());
```
`player_num` is one-based, but the inputDelays array in udpServer is zero-based, which is why it is `player_num - 1`

joeldenning commented:
I've updated this with your feedback, and also expanded it to allow for clients to specify their desired input delay when creating a room or joining a room.

loganmc10 commented:
I still see one issue, inside sendInput:
```
checkIfExists(playerNum, count);
qToBigEndian(inputs[playerNum].value(count).first, &buffer[curr]);
```
There is still the possibility that this will be called when the data isn't there (someone requests the inputs for player 1, count 123, but player 1 hasn't sent those inputs yet). sendInput should just return without doing anything in this case, if you leave it as-is, it's going to send back 0's.

```inputs[playerNum].value(count).first``` will be 0 by default (this is what happens when you call .value(count), and it doesn't exist in the list.

My advice would be to modify ```checkIfExists(playerNum, count);``` for your case, and have it return 1 if the value exists, 0 if it doesn't, and have sendInput return early based on that feedback. Hopefully that makes sense.

loganmc10 commented:
The other overall issue that I just thought about (which you probably didn't experience playing locally): The client never resends inputs. It just sends a single UDP packet for each frame.

The way it's written right now, if one of those packets gets lost/dropped, the game will hang forever, since it won't know the inputs for that player for that frame. Does this make sense? You'll need some method to either tell the client to retransmit if it's missing some data, or have some logic like this:

```
server receives inputs for frame 123
server receives inputs for frame 125
server realizes it never got inputs for frame 124, server copies inputs from frame 125 into frame 124
```

joeldenning commented:
> My advice would be to modify checkIfExists(playerNum, count); for your case, and have it return 1 if the value exists, 0 if it doesn't, and have sendInput return early based on that feedback. Hopefully that makes sense.

Yes, makes sense. I have implemented this.

> The way it's written right now, if one of those packets gets lost/dropped, the game will hang forever, since it won't know the inputs for that player for that frame. Does this make sense? You'll need some method to either tell the client to retransmit if it's missing some data, or have some logic like this:

Yes, that makes sense. I implemented your suggestion of copying newer inputs to previous frames. Perhaps over time it might make sense to change that so that the client retransmits.

loganmc10 commented:
Alright I've reviewed this and it looks good. I didn't test out the input delay or anything, I've recently started a new job so I don't have much free time.

So I mostly verified to make sure that it's not going to negativity affect the existing (floating delay) code, and it doesn't seem like it would make any difference there.

So I'm OK with merging this if you're happy with it, just let me know.

joeldenning commented:
Yes, let's merge it. I've done a fair amount of testing with it and think it's in a pretty good spot. Once the corresponding mupen64plus-gui code is merged, I'm planning on testing it out further with a couple people, with the understanding that it's new and might still have kinks.

loganmc10 commented:
Yes no worries, I'll trust you to sort of "own" this feature and improve upon it :)
